docs/rfcs/0004-rfc-connectiomanager.md

```diff
@@ -0,0 +1,80 @@
+
+The Connection Health, an existing component of our codebase, is tasked with sending periodic pings and determining the stability of the VPN based on the responses received (see `ConnectionHealth::healthCheckup()`). It classifies the VPN stability as `Stable`, `Unstable` (some missing responses), or `NoSignal` (no responses received). When the VPN experiences issues with network connectivity or server availability, it naturally transitions into the No Signal state and, ideally, recovers once the issue is resolved. We can leverage this logic to recover from a probe failure within the Connection Manager.
+
+> For specific details about how these failures are handled and communicated with the user, please refer to this [Figma document](https://www.figma.com/file/GZKigVZbnCisMJNU3Hc9xM/Connection-error-messaging?type=design&node-id=26-215&mode=design&t=kfYnUKuXQ9nsGDRR-0).
+
+## Implementation Plan
+
+```c++
+enum ConnectionManagerStates {
+  StateInternetProbe,
+  StateFirewall,
+  StateServerProbe,
+  StateCheckSubscription,
+  StateIdle,
+}
+```
+### Upon VPN Activation (`StateOff` to `StateConnecting`)
+Perform a single cycle going through all the Connection Manager probes.
+- If a probe fails:
+  - Set Controller to `StateOff`
+  - Stop subsequent probes from running
+  - emit a signal to the frontend communicating which probe failed
+- If all probes pass:
+  - Activate VPN (Controller `StateOn`)
+
+### VPN is already active (`StateOn`)
+Perform periodic checks cycling through all the Connection Manager probes.
+- If a probe fails:
+  - Controller continues to stay in `StateOn`. We will __never__ toggle the VPN off without the user explicitly doing so themselves
+  - We will __not__ manipulate the Connection Health stability based on Connection Manager probes
+  - Stop subsequent probes from running
+  - emit a signal to the frontend communicating which probe failed
+- If all probes pass:
+  - Do nothing. VPN stays on and stable
+
+> Note: The second scenario where a check fails when the VPN is already active still needs some more thinking. There will be a follow up design doc on how we will approach that later on.
+
+### How to recover after a failed probe
+- If VPN is not in `StateOn` yet:
+  - Connection Manager will emit signal about the failed probe to notify the frontend
+- If VPN is in `StateOn`:  
+  - Wait until Connection Health stability goes back to Stable or Unstable
+  - Restart the sequence of ConnectionManager checks
+
+![Diagram to explain how Controller, Connection Manager and Connection Health states interact](./images/CM_diagram.png)
```

Comment:
> Diagram

Could the diagram show how ConnectionHealth communicates with ConnectionManager?

Reply:
There is an arrow from Controller to Connection Diagnostics in the diagram that has the label ConnectionHealth::NoSignal, indicating that if connection health stability goes into no signal then we will kick off the diagnostics. Do you have a suggestion for how to show that better?